### PR TITLE
[DO NOT MERGE] Plugin working branch

### PR DESCRIPTION
**Pull request by cliffcolvin**
## What does this PR change?
* Adds new plugin support for opencost
* Adds new datadog plugin for opencost

## Does this PR relate to any other PRs?
* 

## How will this PR impact users?
* 

## Does this PR address any GitHub or Zendesk issues?
* Closes ...

## How was this PR tested?
* 

## Does this PR require changes to documentation?
* 

## Have you labeled this PR and its corresponding Issue as "next release" if it should be part of the next OpenCost release? If not, why not?
*

### CODE REVIEW

File: generate.sh [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env sh
```
Comment (ameijer) [inline generate.sh:1]: let's build a CI job that runs this, to ensure we don't edit generated objects

File: justfile [hunk 2]
```diff
@@ -25,15 +25,17 @@ build-binary VERSION=version:
...
+           -X github.com/opencost/opencost/pkg/version.GitCommit={{commit}}" \
         -o ./costmodel-amd64
 
     cd ./cmd/costmodel && \
         {{commonenv}} GOOS=linux GOARCH=arm64 go build \
         -ldflags \
-          "-X github.com/opencost/opencost/core/pkg/version.Version={{VERSION}} \
-           -X github.com/opencost/opencost/core/pkg/version.GitCommit={{commit}}" \
+          "-X github.com/opencost/opencost/pkg/version.Version={{VERSION}} \
+           -X github.com/opencost/opencost/pkg/version.Architecture=arm64 \
```
Comment (ameijer) [inline justfile]: we can revert these changes

File: protos/customcost/messages.proto [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+  // spec attributes
+  optional CustomCostExtendedAttributes extended_attributes = 16;
+
+}
+
+message CustomCostExtendedAttributes {
+  // FOCUS billing period start
+  optional google.protobuf.Timestamp billing_period_start = 1;
+  // FOCUS billing period end
+  optional google.protobuf.Timestamp billing_period_end = 2;
```
Comment (ameijer) [inline protos/customcost/messages.proto:105]: need to remove all windows from arch diag

File: generate.sh (hunk 1 above)
Comment (nik-kc) [reply generate.sh:1]: Thoughts on switching this to use `go generate`?

File: justfile (hunk 2 above)
Comment (nik-kc) [reply justfile]: Agreed, we don't need `version.Architecture` anymore.

File: pkg/costmodel/router.go [hunk 4]
```diff
@@ -1828,6 +1846,9 @@ func Initialize(additionalConfigWatchers ...*watcher.ConfigMapWatcher) *Accesses
 	a.Router.GET("/cloud/config/disable", a.CloudConfigController.GetDisableConfigHandler())
 	a.Router.GET("/cloud/config/delete", a.CloudConfigController.GetDeleteConfigHandler())
 
+	a.Router.GET("/customCost/total", a.CustomCostQueryService.GetCustomCostTotalHandler())
+	a.Router.GET("/customCost/timeseries", a.CustomCostQueryService.GetCustomCostTimeseriesHandler())
```
Comment (nik-kc) [inline pkg/costmodel/router.go]: These should move into the `if env.IsCustomCostEnabled()` block.

File: pkg/customcost/ingestor.go [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+	}
+
+	// connect the client
+	rpcClient, err := pluginClient.Client()
+	if err != nil {
+		log.Errorf("error connecting client for plugin %s: %v", domain, err)
+		return
+	}
+
+	// TODO HAVE PLUG IN RETURN DATA AS PROTOBUF
```
Comment (nik-kc) [inline pkg/customcost/ingestor.go]: Is this `TODO` finished?

File: pkg/customcost/memoryrepository.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+	domainData, ok := m.data[domain]
+	if !ok {
+		return nil, nil
+	}
+
+	b, ook := domainData[startTime.UTC()]
+	if !ook {
+		return nil, nil
+	}
```
Comment (nik-kc) [inline pkg/customcost/memoryrepository.go]: We're using a `nil` response combined with a `nil` error to represent "not found in repo". Is this a standard pattern?

File: pkg/customcost/memoryrepository.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+// Expire deletes all items in the map with a start time before the given limit
+func (m *MemoryRepository) Expire(limit time.Time) error {
+	m.rwLock.Lock()
+	defer m.rwLock.Unlock()
+
+	for key, integration := range m.data {
+		for startTime := range integration {
+			if startTime.Before(limit) {
+				delete(integration, startTime)
```
Comment (nik-kc) [inline pkg/customcost/memoryrepository.go:105]: Is it safe to modify a map while iterating on it?

File: pkg/customcost/ingestor.go [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+	go ing.build(rebuild)
+
+	go ing.run()
+
+}
+
+func (ing *CustomCostIngestor) Stop() {
+	// If already stopping, log that and return.
+	if !ing.isStopping.CompareAndSwap(false, true) {
+		log.Infof("CloudCost: ingestor: is already stopping")
```
Comment (nik-kc) [inline pkg/customcost/ingestor.go]: `CloudCost` -> `CustomCost`

File: pkg/customcost/pipelineservice.go [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+	ocplugin "github.com/opencost/opencost/core/pkg/plugin"
+	proto "github.com/opencost/opencost/core/pkg/protocol"
+	"github.com/opencost/opencost/core/pkg/util/timeutil"
+)
+
+var protocol = proto.HTTP()
+
+const execFmt = `%s/%s.ocplugin.%s.%s`
+
+// PipelineService exposes CloudCost pipeline controls and diagnostics endpoints
```
Comment (nik-kc) [inline pkg/customcost/pipelineservice.go]: `CloudCost` -> `CustomCost`

File: pkg/customcost/repository.go [hunk 10]
```diff
@@ -0,0 +1,16 @@
+package customcost
+
+import (
+	"time"
+
+	"github.com/opencost/opencost/core/pkg/model/pb"
+)
+
+// Repository is an interface for storing and retrieving CloudCost data
```
Comment (nik-kc) [inline pkg/customcost/repository.go]: `CloudCost` -> `CustomCost`

File: pkg/customcost/status.go [hunk 11]
```diff
@@ -0,0 +1,27 @@
+package customcost
+
+import (
+	"time"
+
+	"github.com/opencost/opencost/core/pkg/opencost"
+	cloudconfig "github.com/opencost/opencost/pkg/cloud"
+)
+
+// Status gives the details and metadata of a CloudCost integration
```
Comment (nik-kc) [inline pkg/customcost/status.go]: `CloudCost` -> `CustomCost`

File: generate.sh (hunk 1 above)
Comment (ameijer) [reply generate.sh:1]: I personally like the idea of the script, as it lets us quickly build a CI job like I did here: https://github.com/opencost/opencost/pull/2625/files#diff-bd4e577d8cce18da4cecc900fca127f9a0d57d706c9cdb3507b778c1d574a54fR14 that makes sure people don't edit the generated files

File: justfile (hunk 2 above)
Comment (ameijer) [reply justfile]: gone

File: generate.sh (hunk 1 above)
Comment (nik-kc) [reply generate.sh:1]: That's valid, I like that solution.

File: pkg/costmodel/router.go (hunk 4 above)
Comment (ameijer) [reply pkg/costmodel/router.go]: good looks, fixed

File: protos/customcost/messages.proto (hunk 3 above)
Comment (ameijer) [reply protos/customcost/messages.proto:105]: done

File: pkg/customcost/ingestor.go (hunk 5 above)
Comment (ameijer) [reply pkg/customcost/ingestor.go]: sure is

File: pkg/customcost/memoryrepository.go (hunk 6 above)
Comment (ameijer) [reply pkg/customcost/memoryrepository.go]: probably not, let's return an empty struct, I will do that now

File: pkg/customcost/memoryrepository.go (hunk 7 above)
Comment (ameijer) [reply pkg/customcost/memoryrepository.go:105]: looks safe: https://stackoverflow.com/a/23230406

File: pkg/customcost/ingestor.go (hunk 8 above)
Comment (ameijer) [reply pkg/customcost/ingestor.go]: fixed

File: pkg/customcost/pipelineservice.go (hunk 9 above)
Comment (ameijer) [reply pkg/customcost/pipelineservice.go]: fixed

File: pkg/customcost/repository.go (hunk 10 above)
Comment (ameijer) [reply pkg/customcost/repository.go]: fixed

File: pkg/customcost/status.go (hunk 11 above)
Comment (ameijer) [reply pkg/customcost/status.go]: fixed

**Comment by ameijer** [comment]
@nik-kc , PRs fixes are here https://github.com/opencost/opencost/pull/2625 , can you take a look?

**Review by teevans — APPROVED** [review]
